src/main/scala/org/broadinstitute/dsde/firecloud/service/TSVFileSupport.scala

```diff
@@ -179,31 +179,74 @@ trait TSVFileSupport {
-                  case Success(ref) => ref
-                  case Failure(_)   => AttributeString(value)
-                }
-            }
-        }
+    // if this value starts and ends with a quote, it should always be treated as a string
+    if (value.startsWith("\"") && value.endsWith("\"")) {
+      AttributeString(value.substring(1, value.length - 1))
+    } else {
+      // else, inspect the value to find an appropriate datatype
+      toAttribute(value)
+    }
+
+  // the `toAttribute` method checks the first non-whitespace character of inbound TSV cells.
+  private val possibleNumbers = "0123456789+-." // possible first characters for a number
+  private val possibleBooleans = "tfTF" // possible first characters for a boolean
+  private val possibleReferences = "{" // possible first characters for an entity reference
+  private val possibleNonStrings = possibleNumbers + possibleBooleans + possibleReferences
+
+  private def toAttribute(value: String): Attribute = {
+    val trimmed = value.trim
+    // empty string
+    if (trimmed.isEmpty) {
+      return AttributeString(value)
+    }
+    // find the first character of the inbound cell
+    val firstChar = trimmed.charAt(0)
+
+    // first char doesn't match any known starters; this is a string
+    if (!possibleNonStrings.contains(firstChar)) {
+      return AttributeString(value)
+    }
+
+    // could this be a number?
+    if (possibleNumbers.contains(firstChar)) {
+      Try(java.lang.Integer.parseInt(value)) match {
+        case Success(intValue) => return AttributeNumber(intValue)
+        case Failure(_) =>
+          Try(java.lang.Double.parseDouble(value)) match {
+            // because we represent AttributeNumber as a BigDecimal, and BigDecimal has no concept of infinity or NaN,
+            // if we find infinite/NaN numbers here, don't save them as AttributeNumber; instead let them fall through
+            // to AttributeString.
+            case Success(doubleValue)
+                if !Double.NegativeInfinity.equals(doubleValue)
+                  && !Double.PositiveInfinity.equals(doubleValue)
+                  && !Double.NaN.equals(doubleValue)
+                  && !matchesLiteral(value) =>
+              return AttributeNumber(doubleValue)
+            case _ => // noop
+          }
+      }
+    }
+
+    // could this be a boolean?
+    if (possibleBooleans.contains(firstChar)) {
+      Try(BooleanUtils.toBoolean(value.toLowerCase, "true", "false")) match {
+        case Success(booleanValue) => return AttributeBoolean(booleanValue)
+        case _                     => // noop
+      }
     }
 
+    // could this be a reference?
+    if (possibleReferences.contains(firstChar)) {
+      Try(value.parseJson.convertTo[AttributeEntityReference]) match {
+        case Success(ref) => return ref
+        case _            => // noop
+      }
+    }
+
+    // it's a string.
+    AttributeString(value)
+  }
```

Comment:
The early exits bother me in scala code so I asked copilot to rewrite using `collectFirst` and this is what it came up with.
```suggestion
  private def toAttribute(value: String): Attribute = {
    val trimmed = value.trim
    if (trimmed.isEmpty) {
      // empty string
      AttributeString(value)
    } else {
      val firstChar = trimmed.charAt(0)

      val possibleConversions = Seq(
        () =>
          // first char doesn't match any known starters; this is a string
          if (!possibleNonStrings.contains(firstChar)) Some(AttributeString(value))
          else None,
        () =>
          // could this be an int?
          if (possibleNumbers.contains(firstChar)) Try(java.lang.Integer.parseInt(value)).toOption.map(AttributeNumber(_))
          else None,
        () =>
          // could this be a double?
          if (possibleNumbers.contains(firstChar))
            Try(java.lang.Double.parseDouble(value)).toOption
              // because we represent AttributeNumber as a BigDecimal, and BigDecimal has no concept of infinity or NaN,
              // if we find infinite/NaN numbers here, don't save them as AttributeNumber; instead let them fall through
              // to AttributeString.
              .filterNot(doubleValue =>
                Double.NegativeInfinity.equals(doubleValue) || Double.PositiveInfinity.equals(doubleValue) || Double.NaN
                  .equals(doubleValue) || matchesLiteral(value)
              )
              .map(AttributeNumber(_))
          else None,
        () =>
          // could this be a boolean?
          if (possibleBooleans.contains(firstChar))
            Try(BooleanUtils.toBoolean(value.toLowerCase, "true", "false")).toOption.map(AttributeBoolean)
          else None,
        () =>
          // could this be a reference?
          if (possibleReferences.contains(firstChar)) Try(value.parseJson.convertTo[AttributeEntityReference]).toOption
          else None
      )

      possibleConversions
        .collectFirst { case conversion if conversion().isDefined => conversion().get }
        .getOrElse(AttributeString(value))
    }
  }
```

Reply:
@dvoet I've rewritten to be more idiomatic - what do you think?